Lib/os.py

```diff
@@ -485,58 +489,81 @@ def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=
     def _fwalk(topfd, toppath, isbytes, topdown, onerror, follow_symlinks):
         # Note: This uses O(depth of the directory tree) file descriptors: if
         # necessary, it can be adapted to only require O(1) FDs, see issue
-        # #13734.
-
-        scandir_it = scandir(topfd)
-        dirs = []
-        nondirs = []
-        entries = None if topdown or follow_symlinks else []
-        for entry in scandir_it:
-            name = entry.name
-            if isbytes:
-                name = fsencode(name)
-            try:
-                if entry.is_dir():
-                    dirs.append(name)
-                    if entries is not None:
-                        entries.append(entry)
+        # bpo-13734 (gh-57943).
+        stack = [(_WalkAction.WALK, (topfd, toppath))]
+        try:
+            while stack:
+                action, value = stack.pop()
+                if action is _WalkAction.YIELD:
+                    yield value
+                    continue
+                elif action is _WalkAction.CLOSE:
+                    close(value)
+                    continue
+                elif action is _WalkAction.WALK:
+                    topfd, toppath = value
                 else:
-                    nondirs.append(name)
-            except OSError:
-                try:
-                    # Add dangling symlinks, ignore disappeared files
-                    if entry.is_symlink():
-                        nondirs.append(name)
-                except OSError:
-                    pass
+                    raise ValueError(f"invalid walk action: {action}")
```

Comment:
A `ValueError` should generally be for bad user input of some kind. This doesn't look like that: we should only get here (if I understand correctly) if there's been some kind of programming error. An assertion might be more idiomatic here:

```suggestion
                    assert False, f"invalid walk action: {action}"
```

Reply:
Sounds good. Yeah, this theoretically should never happen. Maybe just `raise AssertionError` would be more direct than `assert False`? Seems like that's the more common way to do it as well when the condition has already been proven False

Reply:
Some people care very passionately about the difference between `assert False` and `raise AssertionError`, but I am not one of them -- either is fine by me :)